### ISSUE TITLE
`RecursionError` when running `david_map_reduce_weighted.py`

### ISSUE THREAD

**Issue by richardmillson** [gchq/coreax#849]
### What's the problem?

I encounter a `RecursionError: maximum recursion depth exceeded` when running https://github.com/gchq/coreax/blob/main/examples/david_map_reduce_weighted.py . I'm only using CPU and don't have a GPU enabled. https://github.com/gchq/coreax/issues/633 mentions recursion and the `MapReduce` solver that caused the error at https://github.com/gchq/coreax/blob/a227120d0eab179a9fd4ddd1ba2205f60396556b/examples/david_map_reduce_weighted.py#L167 .

### How can we reproduce the issue?

```shell
python3 -m pip install jax
python3 -m pip install "coreax[test] @git+https://github.com/gchq/coreax.git"
python3 examples/david_map_reduce_weighted.py
```

### Python version

3.11

### Package version

0.3.0

### Operating system

Linux

### Other packages

No

### Relevant log output

```shell
> python3 examples/david_map_reduce_weighted.py
INFO:jax._src.xla_bridge:Unable to initialize backend 'rocm': module 'jaxlib.xla_extension' has no attribute 'GpuAllocatorConfig'
INFO:jax._src.xla_bridge:Unable to initialize backend 'tpu': INTERNAL: Failed to open libtpu.so: libtpu.so: cannot open shared object file: No such file or directory
Image dimensions: (215, 180)
Computing coreset...

Traceback (most recent call last):
  File "/home/richardmillson/coreax/examples/david_map_reduce_weighted.py", line 251, in <module>
    main()
  File "/home/richardmillson/coreax/examples/david_map_reduce_weighted.py", line 167, in main
    herding_coreset, _ = eqx.filter_jit(mapped_herding_solver.reduce)(data)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/equinox/_jit.py", line 239, in __call__
    return self._call(False, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/equinox/_module.py", line 1093, in __call__
    return self.__func__(self.__self__, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/equinox/_jit.py", line 212, in _call
    out = self._cached(dynamic_donate, dynamic_nodonate, static)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/solvers/composite.py", line 168, in reduce
    (pre_coreset, output_solver_state, _indices) = _reduce_coreset(dataset)
                                                   ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/solvers/composite.py", line 166, in _reduce_coreset
    return _reduce_coreset(_coreset, final_indices)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/solvers/composite.py", line 166, in _reduce_coreset
    return _reduce_coreset(_coreset, final_indices)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/solvers/composite.py", line 166, in _reduce_coreset
    return _reduce_coreset(_coreset, final_indices)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  [Previous line repeated 876 more times]
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/solvers/composite.py", line 154, in _reduce_coreset
    coreset_ensemble, ensemble_indices = jax.vmap(wrapper)(partitioned_dataset)
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/solvers/composite.py", line 147, in wrapper
    x, _ = self.base_solver.reduce(partition)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/solvers/coresubset.py", line 223, in reduce
    return self.refine(initial_coresubset, solver_state)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/solvers/coresubset.py", line 254, in refine
    gramian_row_mean = self.kernel.gramian_row_mean(x, block_size=bs, unroll=un)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/kernels/base.py", line 412, in gramian_row_mean
    return self.compute_mean(x, x, axis=0, block_size=block_size, unroll=unroll)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/kernels/base.py", line 595, in compute_mean
    accumulated_sum, column_sum_blocks = jax.lax.scan(
                                         ^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/kernels/base.py", line 590, in block_sum
    accumulated_sum, column_sum_slices = jax.lax.scan(
                                         ^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/kernels/base.py", line 586, in slice_sum
    column_sum_slice = jnp.dot(self.compute(x_, y_), w_y)
                               ^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/kernels/base.py", line 158, in compute
    return pairwise(self.compute_elementwise)(x, y)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/util.py", line 195, in pairwise_fn
    return vmap(
           ^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/kernels/scalar_valued.py", line 764, in compute_elementwise
    score_x = self.score_function(x)
              ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/score_matching.py", line 562, in score_function
    gram_matrix_row_means = self.kernel.compute_mean(x_, kde_data, axis=1)
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/kernels/base.py", line 595, in compute_mean
    accumulated_sum, column_sum_blocks = jax.lax.scan(
                                         ^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/kernels/base.py", line 590, in block_sum
    accumulated_sum, column_sum_slices = jax.lax.scan(
                                         ^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/kernels/base.py", line 586, in slice_sum
    column_sum_slice = jnp.dot(self.compute(x_, y_), w_y)
                               ^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/kernels/base.py", line 158, in compute
    return pairwise(self.compute_elementwise)(x, y)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/util.py", line 195, in pairwise_fn
    return vmap(
           ^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/coreax/kernels/scalar_valued.py", line 275, in compute_elementwise
    return self.output_scale * jnp.exp(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/jax/_src/numpy/array_methods.py", line 1050, in op
    return getattr(self.aval, f"_{name}")(self, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/jax/_src/numpy/array_methods.py", line 573, in deferring_binary_op
    return binary_op(*args)
           ^^^^^^^^^^^^^^^^
  File "/home/richardmillson/.local/lib/python3.12/site-packages/jax/_src/numpy/ufunc_api.py", line 177, in __call__
    return call(*args)
           ^^^^^^^^^^^
RecursionError: maximum recursion depth exceeded
--------------------
For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.
```

**Comment by richardmillson** [comment]
I have the same version of `jax` installed as in https://github.com/gchq/coreax/blob/a227120d0eab179a9fd4ddd1ba2205f60396556b/requirements-doc.txt#L36

```shell
> python3 -m pip list | grep jax
jax                               0.4.35
jaxlib                            0.4.35
jaxopt                            0.8.3
jaxtyping                         0.2.34
```

**Comment by qh681248** [comment]
Hi @richardmillson,

Thank you for bringing this to our attention! We’ve updated the example to address this issue in [PR #821](https://github.com/gchq/coreax/pull/821). This update is scheduled to be merged into the main branch as part of the upcoming version 0.3.1 release. Could you check if this resolves the issue on your end and let us know? Your feedback is much appreciated!

Best,
Coreax Team

**Comment by tp832944** [comment]
@richardmillson we've just released 0.3.1. Hopefully this covers your bug reports. Let us know if it doesn't. Thank you for supporting Coreax!

**Comment by richardmillson** [comment]
Thank you @qh681248 and @tp832944 , the example works for me now! Thanks for making this library!
